resource/resource.go

```diff
@@ -230,24 +235,46 @@ func (r *Resource) Out(
 	atc := env.Get("ATC_EXTERNAL_URL")
 	team := env.Get("BUILD_TEAM_NAME")
 	buildN := env.Get("BUILD_NAME")
-	// https://ci.example.com/teams/developers/pipelines/cognito/jobs/hello/builds/25
+	// https://ci.example.com/teams/main/pipelines/cogito/jobs/hello/builds/25
 	targetURL := fmt.Sprintf("%s/teams/%s/pipelines/%s/jobs/%s/builds/%s",
 		atc, team, pipeline, job, buildN)
 	description := "Build " + buildN
-	log.Debugf("Posting state %v, owner %v, repo: %v, ref %v, context %v, target_url %v",
+	log.Debugf(`out: posting:
+  state: %v
+  owner: %v
+  repo: %v
+  ref: %v
+  context: %v
+  targetURL: %v`,
 		state, owner, repo, ref, context, targetURL)
 	err = status.Add(ref, state, targetURL, description)
 	if err != nil {
 		return nil, nil, err
 	}
-	log.Infof("State (%v) posted successfully", state)
+	log.Infof("out: GitHub state %s for ref %s posted successfully", state, ref[0:9])
 
 	metadata := oc.Metadata{}
 	metadata = append(metadata, oc.NameVal{Name: "state", Value: state})
 
 	return dummyVersion, metadata, nil
 }
 
+func stringify1(xs map[string]interface{}) string {
+	var bld strings.Builder
+	for k, v := range xs {
+		fmt.Fprintf(&bld, "  %s: %v\n", k, v)
+	}
+	return bld.String()
+}
+
+func stringify2(xs map[string]string) string {
+	var bld strings.Builder
+	for k, v := range xs {
+		fmt.Fprintf(&bld, "  %s: %v\n", k, v)
+	}
+	return bld.String()
+}
```

Comment:
For my own education, isn't a `string` compatible with `interface{}`? I'm expecting any type to be compatible with it so `stringify1` should be enough in all cases, shouldn't it?

Reply:
This is what I was expecting also :-(
I have to admit I don't have a full explanation.

Reply:
Researching a bit, `string` is compatible with `interface{}`, in the sense that this would work:

```go
	var s string

	func foo (x interface{}) string {
		return fmt.Sprintf("%v", x)
	}

	foo(s)
```

The problem here is that type conversion happens only for "flat" types, and `map[string]string` is not flat.

I could do:

```go
func stringify(x interface{}) string {
	if xs, ok := x.(map[string]interface{}); ok {
		var bld strings.Builder
		for k, v := range xs {
			fmt.Fprintf(&bld, "  %s: %v\n", k, v)
		}
		return bld.String()
	}
	if xs, ok := x.(map[string]string); ok {
		var bld strings.Builder
		for k, v := range xs {
			fmt.Fprintf(&bld, "  %s: %v\n", k, v)
		}
		return bld.String()
	}
	return fmt.Sprintf("  %v\n", x)
}
```

but, besides just being silly, it is even worse, because it would accept at compile time also types that are not maps, while in this case we know it is either `map[string]string` or `map[string]interface{}`.

I might be missing some Go type foo, but this is the less worse I could find.

Reply:
The nested type conversion make sense. Let's wait for the generic to have a clean version. Apparently they missed the 1.17 release...

Reply:
yes, this is one of the few cases where generics would have helped.